src/rust/engine/fs/src/lib.rs

```diff
@@ -165,6 +164,60 @@ impl GitignoreStyleExcludes {
     self.is_ignored_path(stat.path(), is_dir)
   }
 
+  fn gitignore_file_paths(&self, root: &Path, stat: &Stat) -> Vec<PathBuf> {
+    let path_buf = PathBuf::from(root).join(stat.path());
+    let mut path_parent: Option<&Path> = path_buf.parent();
+    let mut gitignore_paths: Vec<PathBuf> = vec![];
+    while let Some(path) = path_parent {
+      gitignore_paths.push(PathBuf::from(path).join(".gitignore"));
+      if path.parent() == root.parent() {
+        break;
+      }
+      path_parent = path.parent();
+    }
+    gitignore_paths.into_iter().rev().collect()
+  }
+
+  fn is_ignored_with_gitignore_files(
+    &self,
+    stat: &Stat,
+    root: &Path,
+    gitignore_files: Vec<(String, PathBuf)>,
+  ) -> bool {
+    let is_dir = match stat {
+      &Stat::Dir(_) => true,
+      _ => false,
+    };
+    let path = stat.path();
+
+    let mut gitignores: Vec<Gitignore> = gitignore_files
+      .into_iter()
+      .map(|(file_text, path)| {
+        let mut builder = GitignoreBuilder::new(&root);
+        for line in file_text.lines() {
```

Comment:
Can use the `GitignoreBuilder::add` method, which adds the entire content of a `.gitignore` file: https://docs.rs/ignore/0.4.12/ignore/gitignore/struct.GitignoreBuilder.html#method.add